test/internal/layers/lazy.go

```diff
@@ -0,0 +1,166 @@
+			return
+		}
+
+		switch x.Platform {
+		case constants.PlatformLinux:
+			err = x.linuxImage(ctx, dir)
+		case constants.PlatformWindows:
+			err = x.windowsImage(ctx, dir)
+		default:
+			err = fmt.Errorf("unsupported platform %q", x.Platform)
+		}
+	})
+	if err != nil {
+		x.Close(ctx)
+		tb.Fatal(err)
+	}
+	return x.layers
+}
+
+// Close removes the downloaded image layers.
+func (x *LazyImageLayers) Close(ctx context.Context) {
+	for _, dir := range x.layers {
+		d, err := filepath.Abs(dir)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("count not get absolute path to %q", dir)
+			continue
+		}
+
+		if _, err := os.Stat(d); err != nil {
+			if !os.IsNotExist(err) {
+				log.G(ctx).WithError(err).Errorf("path %q is not valid", d)
+			}
+			continue
+		}
+		if err := wclayer.DestroyLayer(ctx, d); err != nil {
+			log.G(ctx).WithError(err).Errorf("could not destroy layer %q", d)
+		}
+	}
+}
+
+func (x *LazyImageLayers) linuxImage(ctx context.Context, dir string) error {
+	img, err := crane.Pull(x.Image, crane.WithPlatform(&v1.Platform{OS: x.Platform, Architecture: runtime.GOARCH}))
+	if err != nil {
+		return fmt.Errorf("failed to pull %q: %w", x.Image, err)
+	}
+
+	f, err := os.Create(filepath.Join(dir, "layer.vhd"))
+	if err != nil {
+		return fmt.Errorf("failed to create layer vhd: %w", err)
+	}
+	defer f.Close()
+	// update x.layers so x.close() does the right thing if this function fails
+	x.layers = []string{dir}
+
+	r, w := io.Pipe()
+	eg := errgroup.Group{}
+	eg.Go(func() error {
+		defer w.Close()
+		if err := crane.Export(img, w); err != nil {
+			return fmt.Errorf("export image %q: %w", x.Image, err)
+		}
+		return nil
+	})
+
+	eg.Go(func() error {
+		defer r.Close()
+		if err := tar2ext4.Convert(r, f, tar2ext4.AppendVhdFooter, tar2ext4.ConvertWhiteout); err != nil {
+			return fmt.Errorf("convert image %q to vhd %q: %w", x.Image, f.Name(), err)
+		}
+		if err := f.Sync(); err != nil {
+			return fmt.Errorf("sync vhd %q to disk: %w", f.Name(), err)
+		}
+		f.Close()
+		if err = security.GrantVmGroupAccess(f.Name()); err != nil {
+			return fmt.Errorf("grant vm group access to %s: %w", f.Name(), err)
+		}
+		return nil
+	})
+
+	return eg.Wait()
+}
+
+func (x *LazyImageLayers) windowsImage(ctx context.Context, dir string) error {
+	img, err := crane.Pull(x.Image, crane.WithPlatform(&v1.Platform{OS: x.Platform, Architecture: runtime.GOARCH}))
+	if err != nil {
+		return fmt.Errorf("failed to pull %q: %w", x.Image, err)
+	}
+
+	layers, err := img.Layers()
+	if err != nil {
+		return fmt.Errorf("failed to get image %q layers: %w", x.Image, err)
+	}
+	if err := winio.EnableProcessPrivileges([]string{winio.SeBackupPrivilege, winio.SeRestorePrivilege}); err != nil {
+		return fmt.Errorf("could not set process privileges: %w", err)
+	}
+
+	for i, l := range layers {
+		d := filepath.Join(dir, strconv.FormatInt(int64(i), 10))
+		if err := os.Mkdir(d, 0755); err != nil {
+			return err
+		}
+		rc, err := l.Uncompressed()
```

Comment:
Shouldn't this be the `compressed` layer since `ImportLayerFromTar` below expects the tar stream as input?

Reply:
Compressed would return a `gzip`-ed (or `zstd`-ed) reader, and not the underlying tar, right?

Reply:
I see. The documentation didn't specifically mention anything but looking at [this code](https://github.com/google/go-containerregistry/blob/main/pkg/v1/stream/layer.go#L131) it seems `compressed` returned the gziped version. As long as this works well enough to run our UVM/containers we should be good.

Reply:
I think layers can either be straight tars or tar.gzip, and you have to use the media type? which i am so glad i dont need to deal with right now

Reply:
Yeah te media type should show that.